drivers/i2c/i2c_sam_twi.c

```diff
@@ -0,0 +1,393 @@
+			read_msg_start(twi, &dev_data->msg, addr);
+		} else {
+			write_msg_start(twi, &dev_data->msg, addr);
+		}
+		/* Wait for the transfer to complete */
+		k_sem_take(&dev_data->sem, K_FOREVER);
+
+		if (dev_data->msg.twi_sr > 0) {
+			/* Something went wrong */
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+static void i2c_sam_twi_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct i2c_sam_twi_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct i2c_sam_twi_dev_data *const dev_data = DEV_DATA(dev);
+	Twi *const twi = dev_cfg->regs;
+	struct twi_msg *msg = &dev_data->msg;
+	u32_t isr_status;
+
+	/* Retrieve interrupt status */
+	isr_status = twi->TWI_SR & twi->TWI_IMR;
+
+	/* Not Acknowledged */
+	if (isr_status & TWI_SR_NACK) {
+		msg->twi_sr = isr_status;
+	}
+
+	/* Byte received */
+	if (isr_status & TWI_SR_RXRDY) {
+
+		msg->buf[msg->idx++] = twi->TWI_RHR;
+
+		if (msg->idx == msg->len - 1) {
+			/* Send a STOP condition on the TWI */
+			twi->TWI_CR = TWI_CR_STOP;
+		}
+	}
+
+	/* Byte sent */
+	if (isr_status & TWI_SR_TXRDY) {
+		if (msg->idx == msg->len) {
+			if (msg->flags & I2C_MSG_STOP) {
+				/* Send a STOP condition on the TWI */
+				twi->TWI_CR = TWI_CR_STOP;
+				/* Disable Transmit Ready interrupt */
+				twi->TWI_IDR = TWI_IDR_TXRDY;
+			} else {
+				/* Transfer completed */
+				isr_status |= TWI_SR_TXCOMP;
+			}
+		} else {
+			twi->TWI_THR = msg->buf[msg->idx++];
+		}
+	}
+
+	/* Transfer completed */
+	if (isr_status & TWI_SR_TXCOMP) {
+		/* Disable all enabled interrupts */
+		twi->TWI_IDR = twi->TWI_IMR;
+		/* All data transferred, nothing else to do */
+		k_sem_give(&dev_data->sem);
+	}
+}
+
+static int i2c_sam_twi_initialize(struct device *dev)
+{
+	const struct i2c_sam_twi_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct i2c_sam_twi_dev_data *const dev_data = DEV_DATA(dev);
+	Twi *const twi = dev_cfg->regs;
+	u32_t bitrate_cfg;
+	int ret;
+
+	/* Configure interrupts */
+	dev_cfg->irq_config();
+
+	/* Initialize semaphore */
+	k_sem_init(&dev_data->sem, 0, 1);
+
+	/* Connect pins to the peripheral */
+	soc_gpio_list_configure(dev_cfg->pin_list, dev_cfg->pin_list_size);
+
+	/* Enable module's clock */
+	soc_pmc_peripheral_enable(dev_cfg->periph_id);
+
+	/* Reset TWI module */
+	twi->TWI_CR = TWI_CR_SWRST;
+
+	bitrate_cfg = _i2c_map_dt_bitrate(dev_cfg->bitrate);
+
+	ret = i2c_sam_twi_configure(dev, I2C_MODE_MASTER | bitrate_cfg);
+	if (ret < 0) {
+		SYS_LOG_ERR("Failed to initialize %s device", DEV_NAME(dev));
+		return ret;
+	}
+
+	/* Enable module's IRQ */
+	irq_enable(dev_cfg->irq_id);
+
+	SYS_LOG_INF("Device %s initialized", DEV_NAME(dev));
+
+	return 0;
+}
+
+static const struct i2c_driver_api i2c_sam_twi_driver_api = {
+	.configure = i2c_sam_twi_configure,
+	.transfer = i2c_sam_twi_transfer,
+};
+
+/* I2C0 */
+
+#ifdef CONFIG_I2C_0
+static struct device DEVICE_NAME_GET(i2c0_sam);
+
+static void i2c0_sam_irq_config(void)
+{
+	IRQ_CONNECT(TWI0_IRQn, CONFIG_I2C_0_IRQ_PRI, i2c_sam_twi_isr,
```

Comment:
we should get IRQ data from DTS

Reply:
As I understood there is a plan to move to generate configuration data structures directly from DTS thus the requirement. That sounds reasonable. On the other hand the IRQ, peripheral ID, register addresses data are all available already from header files for all CMSIS compliant arm SoCs. So now we need to re-write this data in DTS format. That's extra work and it's error prone. Maybe there is another way to handle it? Is the config structure generation concept finalized and ready?